File: backend/src/services/graph_service.py
from typing import Optional
from .base import BaseService
from shared.types import GraphData, GraphNode, GraphEdge

class GraphService(BaseService):
    """Service for managing graph data."""
    
    def __init__(self, container):
        super().__init__(container)
        self._graph_data: Optional[GraphData] = None
    
    def on_initialize(self) -> None:
        # Initialize with sample graph
        nodes = [
            GraphNode(id="a", label="Node A"),
            GraphNode(id="b", label="Node B"),
            GraphNode(id="c", label="Node C"),
            GraphNode(id="d", label="Node D"),
            GraphNode(id="e", label="Node E"),
        ]
        edges = [
            GraphEdge(id="e1", source="a", target="b"),
            GraphEdge(id="e2", source="b", target="c"),
            GraphEdge(id="e3", source="c", target="a"),
            GraphEdge(id="e4", source="d", target="c"),
            GraphEdge(id="e5", source="e", target="d"),
        ]
        self._graph_data = GraphData(nodes=nodes, edges=edges)
    
    def get_graph(self) -> Optional[GraphData]:
        return self._graph_data
    
    def add_node(self, id: str, label: str, data: dict = None) -> GraphNode:
        node = GraphNode(id=id, label=label, data=data)
        if self._graph_data:
            self._graph_data.nodes.append(node)
        return node
    
    def add_edge(self, id: str, source: str, target: str, data: dict = None) -> Optional[GraphEdge]:
        if not self._graph_data:
            return None
        
        # Verify nodes exist
        node_ids = {n.id for n in self._graph_data.nodes}
        if source not in node_ids or target not in node_ids:
            return None
        
        edge = GraphEdge(id=id, source=source, target=target, data=data)
        self._graph_data.edges.append(edge)
        return edge
    
    def remove_node(self, node_id: str) -> bool:
        if not self._graph_data:
            return False
        
        # Remove node
        self._graph_data.nodes = [n for n in self._graph_data.nodes if n.id != node_id]
        
        # Remove connected edges
        self._graph_data.edges = [
            e for e in self._graph_data.edges 
            if e.source != node_id and e.target != node_id
        ]
        return True
    
    def remove_edge(self, edge_id: str) -> bool:
        if not self._graph_data:
            return False
        
        self._graph_data.edges = [e for e in self._graph_data.edges if e.id != edge_id]
        return True
    
    def clear(self):
        if self._graph_data:
            self._graph_data.nodes.clear()
            self._graph_data.edges.clear()

File: frontend/src/services/GraphService.js
/**
 * Graph Service - Handles graph data operations
 */

import { BaseService } from './base.js';

export class GraphService extends BaseService {
  constructor(container) {
    super(container);
    this._graphData = {
      nodes: [],
      edges: [],
    };
  }

  onInitialize() {
    // Initialize with sample graph
    this._graphData = {
      nodes: [
        { data: { id: 'a', label: 'Node A' } },
        { data: { id: 'b', label: 'Node B' } },
        { data: { id: 'c', label: 'Node C' } },
        { data: { id: 'd', label: 'Node D' } },
        { data: { id: 'e', label: 'Node E' } },
      ],
      edges: [
        { data: { id: 'e1', source: 'a', target: 'b' } },
        { data: { id: 'e2', source: 'b', target: 'c' } },
        { data: { id: 'e3', source: 'c', target: 'a' } },
        { data: { id: 'e4', source: 'd', target: 'c' } },
        { data: { id: 'e5', source: 'e', target: 'd' } },
      ],
    };
  }

  async getGraph() {
    try {
      const apiClient = this.container.resolve(Symbol.for('ApiClient'));
      const response = await apiClient.get('/graph');
      if (response.success && response.data) {
        this._graphData = {
          nodes: response.data.nodes.map(n => ({ data: n })),
          edges: response.data.edges.map(e => ({ data: e })),
        };
      }
    } catch (error) {
      console.warn('[GraphService] API unavailable, using local data');
    }
    return this._graphData;
  }

  getLocalGraph() {
    return this._graphData;
  }

  addNode(id, label, data = {}) {
    const node = { data: { id, label, ...data } };
    this._graphData.nodes.push(node);
    return node;
  }

  addEdge(id, source, target, data = {}) {
    // Verify nodes exist
    const nodeIds = new Set(this._graphData.nodes.map(n => n.data.id));
    if (!nodeIds.has(source) || !nodeIds.has(target)) {
      console.error('[GraphService] Cannot add edge: nodes not found');
      return null;
    }

    const edge = { data: { id, source, target, ...data } };
    this._graphData.edges.push(edge);
    return edge;
  }

  removeNode(nodeId) {
    this._graphData.nodes = this._graphData.nodes.filter(n => n.data.id !== nodeId);
    this._graphData.edges = this._graphData.edges.filter(
      e => e.data.source !== nodeId && e.data.target !== nodeId
    );
  }

  removeEdge(edgeId) {
    this._graphData.edges = this._graphData.edges.filter(e => e.data.id !== edgeId);
  }

  clear() {
    this._graphData.nodes = [];
    this._graphData.edges = [];
  }

  setGraph(nodes, edges) {
    this._graphData = { nodes, edges };
  }
}

export default GraphService;